docs/design/value-attribution.md

```diff
@@ -0,0 +1,79 @@
+# Value Attribution
+
+### Brandon Iglesias, Dennis Coyle
+
+
+## Definitions
+
+*Attribution* - _Acknowledgment for establishing the relationship between Storj and a customer_
+
+## OVERVIEW
+
+When our partners bring data to the Tardigrade network we want the ability to attach attribution for the data. 
+
+Our partners will have connectors that their customers will use to store data on the Tardigrade network. When one of their customers uses the connector to store data in a bucket on the Tardigrade network, we want to give that partner attribution for the data that is stored in it.
+
+## GOALS
+
+
+* A generic library that can be used in all  Golang third party connectors
+
+* Data attribution on a per bucket basis
+
+* Ability to calculate how much data was stored in the bucket for a given amount of time. 
+
+* Ability to calculate attributed data bandwidth on the Storj Network
+
+
+## NON GOALS
+
+* Attributions per object
+
+* Multiple attributions per bucket
+
+* Attributions per project
+
+* Attributions per satellite account
+
+
+
+## SCENARIOS
+
+
+* A client points partner connector A to a bucket with NO data in it.
+
+	* The partner receives the attribution for all of the data that the client stores and retrieves from it. 
+
+	* The bucket must be empty when the partner connector is connected to it. 
+
+* A client points partner connector A to an existing bucket that NO OTHER connector has pointed to but has data in it. 
+
+	* The partner does NOT receive attribution for data stored or retrieved from that bucket. 
+
+* A client points partner connector A to an existing bucket that partner connector B had already been pointed to. 
+
+	* Partner connector B receives the attribution for all of the data that is stored and retrieved from the bucket. 
+
+## DESIGN
+### Connector
+Each partner will have a registered id, (which we will refer to as the partner id) that will identify a partners connector on the Storj network.  When a user uploads data to a specified bucket through the connector,  the connector will include the partner id in the context of the request. Before an upload occurs, the uplink will communicate the partner id and bucket name with the tardigrade satellite, checking for a previous attribution. If no attribution is found on the specified bucket and the bucket is currently void of data, the satellite will attribute the partners id to that bucket within the metadata struct. Concurrently to updating the metadata struct the satelitte will add the necessary data to the Attribution table. 
```

Comment:
Why isn't the partner id specified during bucket creation instead of this lazily on first write system?

Reply:
because the bucket can be created months before a bucket is associated with a partner

Reply:
as long as the bucket is empty and has no previous attribution, we want to allow a partner to be attributed to that bucket.

Reply:
if the bucket has been created months ago, but this only works if it's empty, why not just make them recreate the bucket?

like, the age of the bucket doesn't matter, what matters is that it's empty, and if it's empty, recreation is trivial.

Reply:
I guess I don't see the benefit of `create bucket x with no attribution, delete bucket x, create bucket x with attribution` vs `create bucket x with no attribution, add attribution`

Reply:
So if I'm understanding correctly, if you have your uplink command set with a partner id and you upload to a bucket with data in it, it will just happily allow it. If that's the case, I can imagine a user attempting to set a partner id on a bucket that isn't empty and getting no feedback that it never worked.

If setting a partner id is an explicit action, there's an opportunity to inform the user when something went wrong. For example, a command like `uplink set-partner-id sj://bucket partner-id` could say "sorry! the bucket has to be empty".

Like, if no buckets existed, is this the design we would use? If so, great. If not, we're doing something different from ideal to handle migrating the case of empty buckets, which seems like a small thing to care about.

Reply:
I've had it explained to me why lazily allocating the bucket to a partner id is the ideal design.

Reply:
@zeebo it would be nice to know why "lazily allocating the bucket to a partner id is the ideal design" for future reference.